### ISSUE TITLE
Replace `file` and `image` widgets with new `media` widget

### ISSUE THREAD

**Issue by erquhart** [netlify/netlify-cms#1687]
The file and image widgets are identical except for a boolean flag that tells the shared component to behave differently for images. We should combine them into a singular `media` widget (as that's the most common verbiage for assets) and pass the flag via field configuration.

**Comment by tomrutgers** [comment]
From a development / maintenance perspective this is a good idea. From an input / user point of view it doesn't feel like the perfect solution to me. Imagine a user looking to add an image. Or a file. Or a video for that matter. It makes more sense to me to have three separate buttons than to have a media button that is somewhat more unclear. Same counts for documentation: If i'm looking for ways to add an image to my config a `media widget` wouldn't be the first thing I'd search for.

**Comment by tomrutgers** [comment]
> This won't actually impact the user at all - the developer can name and label the field however they like, it's just using the media widget behind the scenes.

Yeah, I was looking at it from a UI perspective, but I get now that it can still work jus fine.

> If it's a documentation problem, we should handle it in documentation rather than letting documentation hurdles determine how we structure the project. We should find ways to lead folks to the media widget when they're searching for video/image/file in the docs site.

True, on it. It would help imho to have a `widget: media`,  `type: image` kind of structure though, so even without reading docs (just by looking at a config.yml) it is explicitly clear what a widget does. That would make it more extendable as well in terms of UI on the input side of things in the long run. _(Like special ui for uploading video when you are using `type: video` in the future)_

Long story short: Your explanation helped me put the media widget into  perspective. It feels like a good idea, but let's take it as an opportunity to make any media type more usable instead of making both the file and image widgets more complex by combining the two!

**Comment by erquhart** [comment]
> It would help imho to have a widget: media, type: image kind of structure

Definitely, the media widget will accept configuration for this. By default it will probably work as a file widget. Honestly, the more I think about it, maybe we should combine them into the file widget (which we can adjust for images only in config as discussed), and then release a `media` widget when we have a robust widget that does more than just upload files (that's all the image widget does right now).

> instead of making both the file and image widgets more complex by combining the two!

That's the whole thing though, they're already combined, literally the same component for both, no difference.

**Comment by tomrutgers** [comment]
Sounds good!

**Comment by 2Clutch** [comment]
Hello. Did anybody ever get around to working on this?

**Comment by erezrokah** [comment]
> Hello. Did anybody ever get around to working on this?

Doesn't seem like, up for grabs 😄 Let me know if you want me to assign you to it

**Comment by 2Clutch** [comment]
> Doesn't seem like, up for grabs 😄 Let me know if you want me to assign you to it

Sure thing. I'd like that. Thank you!

**Comment by 2Clutch** [comment]
I'm currently working on setting up the media package and I'm confused about the process regarding CHANGELOG.md. It would seem the creation of a package cause a version bump is that correct? In addition, each changelog entry contains a url. I would like to know more about the origin of said urls. I tried visiting 5 urls or so, chosen randomly, and none was a valid url.

Also, how should I go about creating the `dist` directory generated inside my package? I tried running `npm/yarn/ng build _package_` but nothing worked. I'm thinking there may be a different to go about when dealing with react that I'm probably not aware of. I would love to get some clarifications on this.

**Comment by 2Clutch** [comment]
Hey @erezrokah. May I know why my pr was closed? I saw your feedback and hadn't gotten to it yet. Wouldn't it be better to continue talking there since the feedback would be more targeted to the work being completed?

**Comment by erezrokah** [comment]
Sorry @2Clutch, I though it would be better to specify the new widget api first in the issue, but hadn’t made that clear enough.

Let me re-open the PR

**Comment by andreamussap** [comment]
Hi guys,
in https://axway-open-docs.netlify.app/ project we're also looking for the ability to upload files, specially script samples (`JSON`, `YAML`, `SQL` etc), not only images. 

So, I think that this is the right place for me to add my +1 vote??

![NetlifyCMS_PR1687](https://user-images.githubusercontent.com/10401080/85846401-abe6ce80-b79d-11ea-9180-ed7951c63df9.jpg)

**Comment by erezrokah** [comment]
Hi @andreamussap, this issue is for a separate media widget and not an editor component for the markdown widget. I would recommend adding a new issue for it (we might be able to leverage from `Code Block` to property display the content of script files).

**Comment by 2Clutch** [comment]
Hey @erezrokah! I've been wrestling with this for the last couple of days and I think it might be best if someone else took on this. I'm fairly new to React and figured this would be the best way to get some hands-on experience but this is turning out to be more complex than I anticipated.

I've unassigned myself to the issue. I may come back later to take a crack at another issue. Or maybe event this one if it's still available.

**Comment by erezrokah** [comment]
Thanks for the follow up @2Clutch! You're welcome to try one our other [good first issues](https://github.com/netlify/netlify-cms/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22). They might need better descriptions/implementation guides, so feel free to ask if the issue description is lacking.
